token/program-2022/src/processor.rs

```diff
@@ -412,93 +421,108 @@ impl Processor {
-            let mut mint = Mint::unpack(&account_info.data.borrow())?;
+            account.pack_base();
+        } else if let Ok(mut mint) = StateWithExtensionsMut::<Mint>::unpack(&mut account_data) {
             match authority_type {
                 AuthorityType::MintTokens => {
                     // Once a mint's supply is fixed, it cannot be undone by setting a new
                     // mint_authority
                     let mint_authority = mint
+                        .base
                         .mint_authority
                         .ok_or(Into::<ProgramError>::into(TokenError::FixedSupply))?;
                     Self::validate_owner(
                         program_id,
-                        account_info.key,
                         &mint_authority,
                         authority_info,
+                        authority_info_len,
                         account_info_iter.as_slice(),
                     )?;
-                    mint.mint_authority = new_authority;
+                    mint.base.mint_authority = new_authority;
                 }
                 AuthorityType::FreezeAccount => {
                     // Once a mint's freeze authority is disabled, it cannot be re-enabled by
                     // setting a new freeze_authority
                     let freeze_authority = mint
+                        .base
                         .freeze_authority
                         .ok_or(Into::<ProgramError>::into(TokenError::MintCannotFreeze))?;
                     Self::validate_owner(
                         program_id,
-                        account_info.key,
                         &freeze_authority,
                         authority_info,
+                        authority_info_len,
+                        account_info_iter.as_slice(),
+                    )?;
+                    mint.base.freeze_authority = new_authority;
+                }
+                AuthorityType::CloseAccount => {
+                    let extension = mint.get_extension_mut::<MintCloseAuthority>()?;
+                    let maybe_close_authority: Option<Pubkey> = extension.close_authority.into();
+                    let close_authority =
+                        maybe_close_authority.ok_or(TokenError::AuthorityTypeNotSupported)?;
+                    Self::validate_owner(
+                        program_id,
+                        &close_authority,
+                        authority_info,
+                        authority_info_len,
                         account_info_iter.as_slice(),
                     )?;
-                    mint.freeze_authority = new_authority;
+                    extension.close_authority = new_authority.try_into()?;
                 }
                 _ => {
                     return Err(TokenError::AuthorityTypeNotSupported.into());
                 }
             }
-            Mint::pack(mint, &mut account_info.data.borrow_mut())?;
+            mint.pack_base();
```

Comment:
I guess technically the `pack_base()` is unnecessary for `AuthorityType::CloseAccount` but trying to avoid this is probably more trouble than it's worth

Reply:
that's a good point, I'll move this up for freeze and mint and avoid this call, for sure